File: server/src/main/java/avatar/model/GameData.java
package avatar.model;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

import avatar.Farm.farmItem;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.JSONValue;

import avatar.db.DbManager;
import lombok.Getter;

@Getter
public class GameData {
    private static final GameData instance = new GameData();

    public static final GameData getInstance() {
        return instance;
    }

    private List<ImageInfo> itemImageDatas = new ArrayList<>();
    private List<ImageInfo> farmImageDatas = new ArrayList<>();
    private List<MapItem> mapItems = new ArrayList<>();
    private List<MapItemType> mapItemTypes = new ArrayList<>();
    private List<farmItem> farmItems = new ArrayList<>();


    public void load() {
        loadItemImageData();
        loadFarmImageData();
        loadMapItem();
        loadMapItemType();
        loadItemfarm();
    }


    public void loadItemfarm() {
        farmItems.clear();
        try (Connection connection = DbManager.getInstance().getConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT * FROM `farmitems`;");
             ResultSet rs = ps.executeQuery()) {
            while (rs.next()) {
                int id = rs.getInt("id");
                String name = rs.getString("name");
                int time = rs.getInt("time");
                int quantity = rs.getInt("quantity");
                int sell = rs.getInt("sell");
                farmItems.add(farmItem.builder().id(id).name(name).time(time).quantity(quantity).sell(sell).build());
            }
        } catch (SQLException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    public void loadItemImageData() {
        itemImageDatas.clear();
        try (Connection connection = DbManager.getInstance().getConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT * FROM `avatar_img_data`;");
             ResultSet rs = ps.executeQuery()) {
            while (rs.next()) {
                int id = rs.getInt("item_id");
                int bigImageID = rs.getInt("image_id");
                int x = rs.getInt("x");
                int y = rs.getInt("y");
                int w = rs.getInt("w");
                int h = rs.getInt("h");
                itemImageDatas.add(ImageInfo.builder().id(id).bigImageID(bigImageID).x(x).y(y).w(w).h(h).build());
            }
        } catch (SQLException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    public void loadFarmImageData() {
        farmImageDatas.clear();
        try (Connection connection = DbManager.getInstance().getConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT * FROM `farm_image_data`;");
             ResultSet rs = ps.executeQuery()) {
            while (rs.next()) {
                int id = rs.getInt("id");
                int bigImageID = rs.getInt("image_id");
                int x = rs.getInt("x");
                int y = rs.getInt("y");
                int w = rs.getInt("w");
                int h = rs.getInt("h");
                farmImageDatas.add(
                        ImageInfo.builder().id(id).bigImageID(bigImageID).x(x).y(y).w(w).h(h).build());
            }
        } catch (SQLException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    public void loadMapItem() {
        mapItems.clear();
        try (Connection connection = DbManager.getInstance().getConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT * FROM `map_item`;");
             ResultSet rs = ps.executeQuery()) {
            while (rs.next()) {
                short id = rs.getShort("id");
                short typeID = rs.getShort("type_id");
                byte type = rs.getByte("type");
                byte x = rs.getByte("x");
                byte y = rs.getByte("y");
                mapItems.add(MapItem.builder().id(id).typeID(typeID).type(type).x(x).y(y).build());
            }
        } catch (SQLException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    public void loadMapItemType() {
        mapItemTypes.clear();
        try (Connection connection = DbManager.getInstance().getConnection();
             PreparedStatement ps = connection.prepareStatement("SELECT * FROM `map_item_type`;");
             ResultSet rs = ps.executeQuery()) {
            while (rs.next()) {
                short id = rs.getShort("id");
                String name = rs.getString("name");
                String description = rs.getString("description");
                short imageID = rs.getShort("image");
                short iconID = rs.getShort("icon");
                short priceCoin = rs.getShort("price_coin");
                short priceGold = rs.getShort("price_gold");
                byte buy = rs.getByte("buy");
                byte dx = rs.getByte("dx");
                byte dy = rs.getByte("dy");
                JSONArray jPosition = (JSONArray) JSONValue.parse(rs.getString("position"));
                int size = jPosition.size();
                List<Position> positions = new ArrayList<>();
                for (int i = 0; i < size; i++) {
                    JSONObject obj = (JSONObject) jPosition.get(i);
                    byte x = ((Long) obj.get("x")).byteValue();
                    byte y = ((Long) obj.get("y")).byteValue();
                    Position p = Position.builder()
                            .x(x)
                            .y(y)
                            .build();
                    positions.add(p);
                }
                mapItemTypes.add(MapItemType.builder().id(id).name(name).des(description).imgID(imageID).iconID(iconID)
                        .priceXu(priceCoin).priceLuong(priceGold).buy(buy).dx(dx).dy(dy).listNotTrans(positions)
                        .build());
            }
            rs.close();
            ps.close();
        } catch (SQLException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    public MapItemType findMapItemType(int idType) {
        for (MapItemType mapItemType : mapItemTypes) {
            if (mapItemType.getId() == idType) {
                return mapItemType;
            }
        }
        return null;
    }
}

File: server-node/src/model/GameData.js
/** Port của avatar/model/GameData.java — nạp dữ liệu tĩnh từ DB. */
import { dbManager } from '../db/DbManager.js';
import { ImageInfo } from './ImageInfo.js';
import { MapItem } from './MapItem.js';
import { MapItemType } from './MapItemType.js';
import { Position } from './Position.js';
import { farmItem } from '../farm/farmItem.js';

class GameData {
  constructor() {
    this.itemImageDatas = [];
    this.farmImageDatas = [];
    this.mapItems = [];
    this.mapItemTypes = [];
    this.farmItems = [];
  }

  static getInstance() {
    return gameData;
  }

  getItemImageDatas() { return this.itemImageDatas; }
  getFarmImageDatas() { return this.farmImageDatas; }
  getMapItems() { return this.mapItems; }
  getMapItemTypes() { return this.mapItemTypes; }
  getFarmItems() { return this.farmItems; }

  async load() {
    await this.loadItemImageData();
    await this.loadFarmImageData();
    await this.loadMapItem();
    await this.loadMapItemType();
    await this.loadItemfarm();
  }

  async loadItemfarm() {
    this.farmItems.length = 0;
    try {
      const rows = await dbManager.query('SELECT * FROM `farmitems`;');
      for (const rs of rows) {
        const id = rs.id | 0;
        const name = rs.name;
        const time = rs.time | 0;
        const quantity = rs.quantity | 0;
        const sell = rs.sell | 0;
        this.farmItems.push(farmItem.builder().id(id).name(name).time(time)
          .quantity(quantity).sell(sell).build());
      }
    } catch (e) {
      console.error(e);
    }
  }

  async loadItemImageData() {
    this.itemImageDatas.length = 0;
    try {
      const rows = await dbManager.query('SELECT * FROM `avatar_img_data`;');
      for (const rs of rows) {
        const id = rs.item_id | 0;
        const bigImageID = rs.image_id | 0;
        const x = rs.x | 0;
        const y = rs.y | 0;
        const w = rs.w | 0;
        const h = rs.h | 0;
        this.itemImageDatas.push(
          ImageInfo.builder().id(id).bigImageID(bigImageID).x(x).y(y).w(w).h(h).build());
      }
    } catch (e) {
      console.error(e);
    }
  }

  async loadFarmImageData() {
    this.farmImageDatas.length = 0;
    try {
      const rows = await dbManager.query('SELECT * FROM `farm_image_data`;');
      for (const rs of rows) {
        const id = rs.id | 0;
        const bigImageID = rs.image_id | 0;
        const x = rs.x | 0;
        const y = rs.y | 0;
        const w = rs.w | 0;
        const h = rs.h | 0;
        this.farmImageDatas.push(
          ImageInfo.builder().id(id).bigImageID(bigImageID).x(x).y(y).w(w).h(h).build());
      }
    } catch (e) {
      console.error(e);
    }
  }

  async loadMapItem() {
    this.mapItems.length = 0;
    try {
      const rows = await dbManager.query('SELECT * FROM `map_item`;');
      for (const rs of rows) {
        const id = (rs.id << 16) >> 16;
        const typeID = (rs.type_id << 16) >> 16;
        const type = (rs.type << 24) >> 24;
        const x = (rs.x << 24) >> 24;
        const y = (rs.y << 24) >> 24;
        this.mapItems.push(MapItem.builder().id(id).typeID(typeID).type(type).x(x).y(y).build());
      }
    } catch (e) {
      console.error(e);
    }
  }

  async loadMapItemType() {
    this.mapItemTypes.length = 0;
    try {
      const rows = await dbManager.query('SELECT * FROM `map_item_type`;');
      for (const rs of rows) {
        const id = (rs.id << 16) >> 16;
        const name = rs.name;
        const description = rs.description;
        const imageID = (rs.image << 16) >> 16;
        const iconID = (rs.icon << 16) >> 16;
        const priceCoin = (rs.price_coin << 16) >> 16;
        const priceGold = (rs.price_gold << 16) >> 16;
        const buy = (rs.buy << 24) >> 24;
        const dx = (rs.dx << 24) >> 24;
        const dy = (rs.dy << 24) >> 24;
        const jPosition = JSON.parse(rs.position);
        const size = jPosition.length;
        const positions = [];
        for (let i = 0; i < size; i++) {
          const obj = jPosition[i];
          const x = (obj.x << 24) >> 24;
          const y = (obj.y << 24) >> 24;
          const p = Position.builder()
            .x(x)
            .y(y)
            .build();
          positions.push(p);
        }
        this.mapItemTypes.push(MapItemType.builder().id(id).name(name).des(description)
          .imgID(imageID).iconID(iconID)
          .priceXu(priceCoin).priceLuong(priceGold).buy(buy).dx(dx).dy(dy)
          .listNotTrans(positions)
          .build());
      }
    } catch (e) {
      console.error(e);
    }
  }

  findMapItemType(idType) {
    for (const mapItemType of this.mapItemTypes) {
      if (mapItemType.getId() === idType) {
        return mapItemType;
      }
    }
    return null;
  }
}

export const gameData = new GameData();
export { GameData };
export default gameData;
